core/test/mpi/distributed/solver/multigrid.cpp

```diff
@@ -0,0 +1,212 @@
+                    .with_criteria(
+                        gko::stop::Iteration::build().with_max_iters(1u))
+                    .on(this->ref))
+            .on(this->ref);
+    gko::size_type n = 5;
+    gko::size_type global_n = n * this->comm.size();
+    auto mtx = gko::share(
+        dist_mtx_type::create(this->ref, this->comm, gko::dim<2>{global_n},
+                              DummyLinOp::create(this->ref, gko::dim<2>{n})));
+
+    auto mg = mg_factory->generate(mtx);
+
+    auto mg_level = mg->get_mg_level_list();
+    auto first_n = global_n - this->comm.size();
+    auto second_n = first_n - this->comm.size();
+    ASSERT_EQ(mg_level.at(0)->get_fine_op()->get_size(), gko::dim<2>(global_n));
+    ASSERT_EQ(mg_level.at(0)->get_restrict_op()->get_size(),
+              gko::dim<2>(first_n, global_n));
+    ASSERT_EQ(mg_level.at(0)->get_prolong_op()->get_size(),
+              gko::dim<2>(global_n, first_n));
+    ASSERT_EQ(mg_level.at(0)->get_coarse_op()->get_size(),
+              gko::dim<2>(first_n));
+    ASSERT_EQ(dynamic_cast<
+                  const gko::experimental::distributed::DistributedLocalSize*>(
+                  mg_level.at(0)->get_coarse_op().get())
+                  ->get_local_size(),
+              gko::dim<2>(n - 1));
+    // next mg_level
+    ASSERT_EQ(mg_level.at(1)->get_fine_op()->get_size(), gko::dim<2>(first_n));
+    ASSERT_EQ(mg_level.at(1)->get_restrict_op()->get_size(),
+              gko::dim<2>(second_n, first_n));
+    ASSERT_EQ(mg_level.at(1)->get_prolong_op()->get_size(),
+              gko::dim<2>(first_n, second_n));
+    ASSERT_EQ(mg_level.at(1)->get_coarse_op()->get_size(),
+              gko::dim<2>(second_n));
+}
+
+
+TEST_F(Multigrid, ApplyWithoutException)
+{
+    auto mg_factory =
+        multigrid::build()
+            .with_max_levels(2u)
+            .with_mg_level(mg_level::build())
+            .with_pre_smoother(nullptr)
+            .with_mid_smoother(nullptr)
+            .with_criteria(gko::stop::Iteration::build().with_max_iters(1u))
+            .with_min_coarse_rows(1u)
+            .with_coarsest_solver(
+                gko::solver::Cg<value_type>::build()
+                    .with_criteria(
+                        gko::stop::Iteration::build().with_max_iters(1u))
+                    .on(this->ref))
+            .on(this->ref);
+    gko::size_type n = 5;
+    gko::size_type global_n = n * this->comm.size();
+    auto mtx = gko::share(
+        dist_mtx_type::create(this->ref, this->comm, gko::dim<2>{global_n},
+                              DummyLinOp::create(this->ref, gko::dim<2>{n})));
+    auto b = dist_vec_type::create(this->ref, this->comm,
+                                   gko::dim<2>{global_n, 1}, gko::dim<2>{n, 1});
+    auto x = dist_vec_type::create(this->ref, this->comm,
+                                   gko::dim<2>{global_n, 1}, gko::dim<2>{n, 1});
+
+    auto mg = mg_factory->generate(mtx);
+
+    ASSERT_NO_THROW(mg->apply(b, x));
```

Comment:
I think distributed coarsening method usually different from the non-distributed one?

Reply:
As you are testing the solver here, I think checking if the two solutions are within some tolerance makes sense.

Reply:
ah, for this situation, I can not do it because we do not have actual distributed coarsening method.
I think the current test should be enough because the distributed multigrid actually only handle the vector allocation. These tests ensure the allocation is correct and the corresponding apply works correctly

Reply:
Ah, yes. Maybe make a note/TODO here for now, and it should be okay to add it later